core/src/HIP/Kokkos_HIP_Parallel_MDRange.hpp

```diff
@@ -44,153 +44,370 @@
+      }
+
+      if (Experimental::Impl::HIPTraits::WarpSize < word_count.value) {
+        __syncthreads();
+      }
+
+      for (unsigned i = hipThreadIdx_y; i < word_count.value;
+           i += hipBlockDim_y) {
+        global[i] = shared[i];
+      }
+    }
+  }
+
+  // Determine block size constrained by shared memory:
+  inline unsigned local_block_size(const FunctorType& f) {
+    unsigned n     = Experimental::Impl::HIPTraits::WarpSize * 8;
+    int shmem_size = ::Kokkos::Impl::hip_single_inter_block_reduce_scan_shmem<
+        false, FunctorType, WorkTag>(f, n);
+    while (
+        (n &&
+         (m_policy.space().impl_internal_space_instance()->m_maxShmemPerBlock <
+          shmem_size)) ||
+        (n > static_cast<unsigned>(
+                 ::Kokkos::Experimental::Impl::hip_get_max_block_size<
+                     ParallelReduce, LaunchBounds>(f, 1, shmem_size, 0)))) {
+      n >>= 1;
+      shmem_size = ::Kokkos::Impl::hip_single_inter_block_reduce_scan_shmem<
+          false, FunctorType, WorkTag>(f, n);
+    }
+    return n;
+  }
+
+  inline void execute() {
+    const int nwork = m_policy.m_num_tiles;
+    if (nwork) {
+      int block_size = m_policy.m_prod_tile_dims;
+      // CONSTRAINT: Algorithm requires block_size >= product of tile dimensions
+      // Nearest power of two
+      int exponent_pow_two    = std::ceil(std::log2(block_size));
+      block_size              = std::pow(2, exponent_pow_two);
+      int suggested_blocksize = local_block_size(m_functor);
+
+      block_size = (block_size > suggested_blocksize)
+                       ? block_size
+                       : suggested_blocksize;  // Note: block_size must be less
+                                               // than or equal to 512
+
+      m_scratch_space =
+          ::Kokkos::Experimental::Impl::hip_internal_scratch_space(
+              ValueTraits::value_size(
+                  ReducerConditional::select(m_functor, m_reducer)) *
+              block_size /* block_size == max block_count */);
+      m_scratch_flags =
+          ::Kokkos::Experimental::Impl::hip_internal_scratch_flags(
+              sizeof(size_type));
+
+      // REQUIRED ( 1 , N , 1 )
+      const dim3 block(1, block_size, 1);
+      // Required grid.x <= block.y
+      const dim3 grid(std::min(int(block.y), int(nwork)), 1, 1);
+
+      const int shmem =
+          UseShflReduction
+              ? 0
+              : ::Kokkos::Impl::hip_single_inter_block_reduce_scan_shmem<
+                    false, FunctorType, WorkTag>(m_functor, block.y);
+
+      Kokkos::Experimental::Impl::HIPParallelLaunch<ParallelReduce,
+                                                    LaunchBounds>(
+          *this, grid, block, shmem,
+          m_policy.space().impl_internal_space_instance(),
+          false);  // copy to device and execute
+
+      if (!m_result_ptr_device_accessible) {
+        Experimental::HIP().fence();
+
+        if (m_result_ptr) {
```

Comment:
Yeah it's possible because you can just let Kokkos accumulates the results in a dynamic array by just passing a pointer to `Kokkos::parallel_reduce`. Then Kokkos creates an unmanaged view which is used here.